### PR TITLE
Improve plugins init performances

### PR DESCRIPTION
**Pull request by AdrienClairembault**
I've noticed a while ago that the performances cost of plugins seemed to be a bit too high.
I can't tell if it was an exponential scale or not but the more plugins you had the slower GLPI would get.

I've decided to take a quick look at it today and managed to cut the plugins initialization process time by half.
There may be more optimizations to be found but it's a good first step.

This was done by moving some repetitive code that would be run multiple time with the same input into class properties that are only loaded once.

To benchmark the performances on my side, I've enabled about 20 plugins and added the code below to time the `plugin::init()` function, which is called once per page or ajax request and will load all the plugins.
```diff
diff --git a/src/Plugin.php b/src/Plugin.php
index 5ee1f541b1..d2592d808c 100644
--- a/src/Plugin.php
+++ b/src/Plugin.php
@@ -229,6 +229,7 @@ class Plugin extends CommonDBTM
     public function init(bool $load_plugins = false, array $excluded_plugins = [])
     {
         global $DB;
+        $start_time = microtime(true);

         self::$plugins_state_checked = false;
         self::$activated_plugins = [];
@@ -270,6 +271,9 @@ class Plugin extends CommonDBTM
             // For plugins which require action after all plugin init
             Plugin::doHook(Hooks::POST_INIT);
         }
+
+        $end_time = microtime(true);
+        Toolbox::logDebug($end_time - $start_time);
     }

```

Then I went on the plugin page (could be any other page), pressed F5 and saved the logs.
Each logs seems to contain 4 entries, which would mean we have 1 page being loaded + 3 ajax requests.

First, here are the executions time in `10.0/bugfixes`:

image

The average time seems to be 0.1s second which is quite big as the total load time of the page is about 0.27ms (so plugins init account for almost half my execution time in a localhost environnement). 

And here are the exections time after applying the optimizations found in this PR:

image

We are down to around 0.05s which is a great improvement for instances with a lot of plugins, especially for things like loading dashboards where there may be a 20+ ajax requests at the same time,.

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |

### CODE REVIEW

File: src/Plugin.php [hunk 1]
```diff
@@ -486,10 +492,28 @@ public function checkStates($scan_inactive_and_new_plugins = false, array $exclu
...
         }
 
-       // Prevent duplicated checks
+        // Prevent duplicated checks
         $directories = array_unique($directories);
 
-       // Check all directories from the checklist
+        // Build plugin info data once before iterating to save performances
+        $plugins_informations = [];
+        $plugins_directories = new DirectoryIterator(GLPI_ROOT . '/plugins');
```
Comment (cconard96) [inline src/Plugin.php:500]: Missing support for plugins added from the Marketplace?

**Comment by cedric-anne** [comment]
Do you have Opcache extension installed/activated?

File: src/Plugin.php (hunk 1 above)
Comment (AdrienClairembault) [reply src/Plugin.php:500]: I'm not sure, the original code was only targeting /plugins too so maybe its on purpose ?

**Comment by AdrienClairembault** [comment]
> Do you have Opcache extension installed/activated?

image

**Comment by AdrienClairembault** [comment]
Tests failed because I didn't noticed that `Plugin::checkPluginState()` was public and call be called independently.
Latest commit should fix the issue.

**Comment by AdrienClairembault** [comment]
I've added marketplace support in the latest commit but it would be nice to understand how the original code was able to work with only `/plugins`

**Comment by cconard96** [comment]
> I've added marketplace support in the latest commit but it would be nice to understand how the original code was able to work with only `/plugins` (or maybe it didn't work).

Was the original code from `getNewInfoAndDirBasedOnOldName`?
I think it predates the Marketplace and I'm not entirely sure how replacements are handled for Marketplace plugins (if at all).

File: src/Plugin.php [hunk 2]
```diff
@@ -124,6 +124,13 @@ class Plugin extends CommonDBTM
      */
     private static $loaded_plugins = [];
 
+    /**
+     * Store additional infos for each plugins
+     *
+     * @var array
+     */
+    private ?array $plugins_informations = null;
```
Comment (trasher) [inline src/Plugin.php:132]:
```suggestion
    private ?array $plugins_information = null;
```

"information" is uncountable

Comment (AdrienClairembault) [reply src/Plugin.php:132]: Done.

**Comment by AdrienClairembault** [comment]
Web tests are failing due to some kind of missing plugin warning, need to investigate.

**Review by cconard96 — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Comment by AdrienClairembault** [comment]
`getInformationsFromDirectory` could be renamed too but not in 10/bf since it's a public method.

**Review by trasher — APPROVED** [review]

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Comment by cedric-anne** [comment]
`Plugin::getInformationsFromDirectory()` should probably use information already fetched by your new method.

**Comment by AdrienClairembault** [comment]
> `Plugin::getInformationsFromDirectory()` should probably use information already fetched by your new method.

The new method use `getInformationsFromDirectory` so we might end up with an infinite reference loop ;)

**Review by Rom1-B — APPROVED** [review]

File: src/Plugin.php [hunk 3]
```diff
@@ -500,6 +506,40 @@ public function checkStates($scan_inactive_and_new_plugins = false, array $exclu
         self::$plugins_state_checked = true;
     }
 
+    /**
+     * Build plugin info data once to save performances
```
Comment (cedric-anne) [inline src/Plugin.php:510]:
```suggestion
     * Get information for all plugins found on filesystem.
```

File: src/Plugin.php [hunk 4]
```diff
@@ -500,6 +506,40 @@ public function checkStates($scan_inactive_and_new_plugins = false, array $exclu
...
+     *
+     * @return array
+     */
+    private function getPluginInformation(): array
```
Comment (cedric-anne) [inline src/Plugin.php:514]:
```suggestion
    private function getPluginsInformation(): array
```

File: src/Plugin.php [hunk 5]
```diff
@@ -510,14 +550,12 @@ public function checkStates($scan_inactive_and_new_plugins = false, array $exclu
      */
     public function checkPluginState($plugin_key)
     {
-
         $plugin = new self();
 
-        $informations = $this->getInformationsFromDirectory($plugin_key);
+        $information = $this->getPluginInformation()[$plugin_key];
```
Comment (cedric-anne) [inline src/Plugin.php:555]: A plugin can be present in DB but not on filesystem.

```suggestion
        $information = $this->getPluginsInformation()[$plugin_key] ?? [];
```

File: src/Plugin.php [hunk 6]
```diff
@@ -728,27 +766,16 @@ public function checkPluginState($plugin_key)
...
-            if (
-                in_array($plugin_directory->getFilename(), ['.svn', '.', '..'])
-                || !is_dir($plugin_directory->getRealPath())
-            ) {
-                continue;
-            }
-
-            $informations = $this->getInformationsFromDirectory($plugin_directory->getFilename());
-            if (array_key_exists('oldname', $informations) && $informations['oldname'] === $oldname) {
+        foreach ($this->plugins_information as $plugin_name => $information) {
```
Comment (cedric-anne) [inline src/Plugin.php:773]:
```suggestion
        foreach ($this->getPluginsInformation() as $plugin_name => $information) {
```

**Review by orthagh — APPROVED** [review]

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**